File: ruby/allergies/allergies.rb
class Allergies

  VALUES = {
    'eggs'         => 1,
    'peanuts'      => 1<<1,
    'shellfish'    => 1<<2,
    'strawberries' => 1<<3,
    'tomatoes'     => 1<<4,
    'chocolate'    => 1<<5,
    'pollen'       => 1<<6,
    'cats'         => 1<<7,
  }

  def initialize(score)
    @score = score
  end

  def allergic_to?(item)
    @score & VALUES[item] != 0
  end

  def list
    VALUES.keys.select { |item| allergic_to?(item) }
  end

end

File: elixir/allergies/allergies.exs
defmodule Allergies do

  use Bitwise

  @allergy_flags [
    {"cats",       128},
    {"pollen",      64},
    {"chocolate",   32},
    {"tomatoes",    16},
    {"strawberries", 8},
    {"shellfish",    4},
    {"peanuts",      2},
    {"eggs",         1},
  ]

  @allergies Enum.into(@allergy_flags, HashDict.new)

  @doc """
  List the allergies for which the corresponding flag bit is true.

  Allergies should be ordered, starting with the allergie with the least
  significant bit ("eggs").
  """
  @spec list(non_neg_integer) :: [String.t]
  def list(flags) do
    Enum.reduce @allergy_flags, [], fn
      ({allergy, flag}, allergies)
        when (flags &&& flag) != 0 -> [allergy|allergies]
      (foo, allergies) -> allergies
    end
  end

  @doc """
  Returns whether the corresponding flag bit in 'flags' is set for the item.
  """
  @spec allergic_to?(non_neg_integer, String.t) :: boolean
  def allergic_to?(flags, item) do
    (flags &&& @allergies[item]) != 0
  end
end
